### ISSUE TITLE
suggestion

### ISSUE THREAD

**Issue by ericlangedijk** [domino14/macondo#375]
I looked at the code of the Gordon move generation and did a little test in my own engine.
1) Your code starts with all childnodes and then looks if the letter is in the rack.
2) My code starts with the rack and then tries to find the childnode.

I tried option #1 to check what the performance did. It got down by almost 1 million moves per second.
Maybe you can try option #2 to check out the difference.

I think option 1 is faster because on average the number of childnodes is bigger than the tiles on the rack.

**Comment by domino14** [comment]
sorry, i didn't quite understand the suggestion - what do you mean by it got down?

**Comment by ericlangedijk** [comment]
I mean the performance.  Normally (in some testpositions) I measure 4.8 million moves per second. Using the "option 1" it decreased to 3.5 million.
So if you turn around the logic in the recursiveGen function maybe you can gain performance.

**Comment by andy-k** [comment]
we aren't doing [find_node](https://github.com/ericlangedijk/scrabble/blob/712b8561838b736a34d0e305ef471512fedccfa4/src/movgen.zig#L238) with [popCount](https://github.com/ericlangedijk/scrabble/blob/712b8561838b736a34d0e305ef471512fedccfa4/src/gaddag.zig#L336-L342), because we aren't storing [unaligned child_ptr and mask](https://github.com/ericlangedijk/scrabble/blob/712b8561838b736a34d0e305ef471512fedccfa4/src/gaddag.zig#L624-L627) in our Node.

we aren't indexing the childnodes, so we can't start with the rack in [recursiveGen](https://github.com/domino14/macondo/blob/6cee6b27b84a069c4795c1ea98433a8fb43504e9/movegen/movegen.go#L286). however, checking the contiguous nodes incurs minimal cache misses.

can you suggest a way to turn around the logic in the recursiveGen function without significantly changing the data structure?

**Comment by ericlangedijk** [comment]
I could not find the actual implementation of the gaddag tree in macondo. And I am not a Go programmer.
But: when childnodes of the 'currrent node' are sequentially in memory it should work with the same amount of cache hits / misses I believe. I build the tree in a flat memory structure, assuring all childnodes are sequential in memory.

so instead of:
`foreach(childnode) check_if_rack_contains_this_letter(childnode.letter)`
it should be possible to do:
`foreach(rackletter) check_if_gaddag_node_exists(rackletter)`

It alll depends on the memory structure. 
I must say this prefetch improved performance as well:
https://github.com/ericlangedijk/scrabble/blob/712b8561838b736a34d0e305ef471512fedccfa4/src/movgen.zig#L229

In general I think there are - on average - more childnodes than rackletters.
Cache hits or not: you still have to loop through 15 children instead of 7 or less.

**Comment by ericlangedijk** [comment]
The code on github is very very far from finished. It just is a movegenerator now. The code is there now mainly for backup purposes.
31 letters will probably stay the maximum in my code. 32 bits are just too convenient :)

The gaddag takes around 1.5 seconds to build from a textfile. And I did not do any optimization regarding size or speed. Is around 120 MB for 360.000 Dutch Words.

The code compiles. You just need a textfile with an english word on each line. And I hardcoded some path to that textfile in the main function.
https://github.com/ericlangedijk/scrabble/blob/712b8561838b736a34d0e305ef471512fedccfa4/src/main.zig#L31
Or did I forget a file? Cannot see it from the message you sent.

How on earth do you put a gaddag structure in 4.8 MB?
I will have a look at your links.

I already changed that trymask stuff into something better last week by using an 'engine rack interally'.

And: of course I am curious how fast macondo generates its moves, lets say on a board with 2 or 3 moves on it and a good rack including 2 blanks.

And still curious if my suggestion would work!

**Comment by andy-k** [comment]
32 bits are indeed convenient! but, we need the bigger alphabet support [here](https://woogles.io/).

i think most affordable computers do not have 120 MB of L3 cache. i don't have a copy of the Dutch Words, so i can't compare directly. CSW19 only contains 279496 words.

zig's `note: while parsing` was really followed by nothing, and that was compile time, so i never reached runtime where changing the hardcoded path would matter.

feel free to pick up Go to explore how fast macondo is and how much you can improve it with your suggestion. we are already aware macondo is not fast, but it's expected, mainly because it's written in Go :)

**Comment by ericlangedijk** [comment]
Okidoki. I read also there is a macondo clone in the making in C. Zig would be a much better idea I think. But maybe that's personal :)

But tell me: you need 5 MB in memory for all these prefixes and suffixes? If so, I will have to study!
Or is it the base-file you are building from?

In the near future I will refresh that github including some wordlists and make sure it compiles.

Speed is not everything. I heard macondo is the strongest. That's worth a lot more.

Diving into Go...I am in the process of learning zig. That would be too much. But I certainly will check some links you gave me. And once in a while look around here. Very interesting project.
There is a lot you can learn from the Stockfish chess engine by the way!

**Comment by domino14** [comment]
@ericlangedijk yes, we have this as well: https://github.com/jvc56/MAGPIE - macondo clone in C, very much WIP but it's already significantly faster than macondo. I am thinking in the next year or two I might actually sunset Macondo and focus all my contributions on MAGPIE. 

the entire gaddag is just a few MB. You can read about @andy-k 's data structure here: https://github.com/andy-k/wolges/blob/main/details.txt for how he was able to get it so small. I can't imagine it would be that much larger for Dutch.

**Comment by domino14** [comment]
@ericlangedijk btw if @jvc56 hadn't started MAGPIE I would definitely have rewritten Macondo in Zig. I still want to play around with that language for something.

**Comment by ericlangedijk** [comment]
Zig is fantastic. I started this year with a traumatic Rust experience for a few months. Don't follow that path.
I am currently a bit 'programming tired' due to my work. The next things I am going to do is stuffing gaddag into 5 or 6 MB. And start with an endgame engine. Maybe we can - somewhere - discuss things around scrabbleprogramming in the near future. I would like that. The first thing I have to solve is a decent game 'history': small, fast and covering all possibilities.

**Comment by domino14** [comment]
Macondo has a fully exhaustive endgame engine with a lot of optimization ideas from chess. I'd love to see improvements to it.

**Comment by ericlangedijk** [comment]
I did quite some chess with zobrist, iterative deepening etc. If I can read the code I will check whats happening there.

**Comment by ericlangedijk** [comment]
Sorry to interrupt here, but I have some burning questions...
Andy, I see you are an trie-expert. Could I ask you some questions? Here or by email?
